src/profiler.ts

```diff
@@ -13,57 +13,203 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+import * as delay from 'delay';
+import * as http from 'http';
 import * as path from 'path';
+import * as zlib from 'zlib';
 
-import {AuthenticationConfig, Common, Logger, Service, ServiceConfig} from '../third_party/types/common-types';
+import {perftools} from '../src/profile';
+import {AuthenticationConfig, Common, Logger, Service, ServiceConfig, ServiceObject, ServiceObjectConfig} from '../third_party/types/common-types';
 
-import {Config} from './config';
+import {ProfilerConfig} from './config';
 import {HeapProfiler} from './profilers/heap-profiler';
 import {TimeProfiler} from './profilers/time-profiler';
 
-const pjson = require('../../package.json');
 const common: Common = require('@google-cloud/common');
+const pjson = require('../../package.json');
+const API = 'https://cloudprofiler.googleapis.com/v2';
 
-export interface ProfilerConfig extends AuthenticationConfig {
-  projectId: string;
-  logLevel: number;
-  serviceContext: {service: string; version?: string;};
-  instance: string;
-  zone: string;
-  disableTime: boolean;
-  disableHeap: boolean;
+// Interface for body of response from Stackdriver Profiler API when creating
+// profile and used as body of request to Stackdriver Profiler API when
+// uploading a profile.
+// Public for testing.
+export interface RequestProfile {
+  name?: string;
+  profileType?: string;
+  duration?: any;
+  profileBytes?: string;
+  labels?: {instance?: string; zone?: string};
 }
 
-// TODO: finish implementing Profiler.
-// TODO: add stop() method to stop profiling.
+// Converts a profile to a compressed, base64 encoded string.
+function profileBytes(p: perftools.profiles.IProfile): Promise<string> {
+  let time = new Date((p.timeNanos as number) / 1000000);
+  let pwriter = perftools.profiles.Profile.encode(p);
+  let buffer = new Buffer(pwriter.finish());
+  return new Promise<string>((resolve, reject) => {
+    zlib.gzip(buffer, (err: Error, gzBuf: Buffer) => {
+      err ? reject(err) : resolve(gzBuf.toString('base64'));
+    });
+  });
+}
+
+// Service object for communicating with Stackdriver Profiler API.
+class ProfilerServiceObject extends common.ServiceObject {
+  constructor(options: AuthenticationConfig) {
+    options = common.util.normalizeArguments(null, options);
+    const config = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(config, options), baseUrl: '/'});
+  }
+}
+
+// Polls Stackdriver Profiler server for instructions on behalf of a task and
+// collects and uploads profiles as requested.
 export class Profiler {
-  config: ProfilerConfig;
-  logger: Logger;
-  service: Service;
+  private config: ProfilerConfig;
```

Comment:
While you're making this change; remove this from here, and instead mark it as a private property in the constructor argument list. This will also get rid of need for the `this.config = ` assignment.